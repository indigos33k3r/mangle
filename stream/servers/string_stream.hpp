#ifndef MANGLE_STREAM_STRINGSERVER_H
#define MANGLE_STREAM_STRINGSERVER_H

#include "memory_stream.hpp"

namespace Mangle {
namespace Stream {

/** A specialized MemoryStream that takes an std::string as input. A
    local copy of the string is stored in the instance.

    NOTE on using MemoryStream::clone(): Clones are invalidated once
    the original StringStream goes out of scope and is deallocated.
 */
class StringStream : public MemoryStream
{
  const std::string str;

 public:
  StringStream(const std::string &_str)
    : str(_str)
  {
    // Call this after initializing the class variable str, since we
    // need the locally allocated pointer.
    set(str.c_str(), str.length());
  }

  static MemoryStreamPtr Open(const std::string &str)
  { return MemoryStreamPtr(new StringStream(str)); }
};

}} // namespaces
#endif
