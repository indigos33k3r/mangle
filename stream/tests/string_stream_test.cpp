#include "../servers/string_stream.hpp"

#include <iostream>
#include <string.h>

using namespace Mangle::Stream;
using namespace std;

int main()
{
  StreamPtr inp = StringStream::Open("hello world!");

  cout << "Size: " << inp->size() << endl;
  cout << "Pos: " << inp->tell() << "\nSeeking...\n";
  inp->seek(3);
  cout << "Pos: " << inp->tell() << endl;
  char data[12];
  memset(data, 0, 12);
  cout << "Reading: " << inp->read(data, 4) << endl;
  cout << "Four bytes: " << data << endl;
  cout << "Eof: " << inp->eof() << endl;
  cout << "Pos: " << inp->tell() << "\nSeeking again...\n";
  inp->seek(33);
  cout << "Pos: " << inp->tell() << endl;
  cout << "Eof: " << inp->eof() << "\nSeek to 6\n";
  inp->seek(6);
  cout << "Eof: " << inp->eof() << endl;
  cout << "Pos: " << inp->tell() << endl;
  cout << "Over-reading: " << inp->read(data, 200) << endl;
  cout << "Result: " << data << endl;
  cout << "Eof: " << inp->eof() << endl;
  cout << "Pos: " << inp->tell() << endl;
  inp->seek(0);
  cout << "Finally, reading the entire string: " << inp->read(data,11) << endl;
  cout << "Result: " << data << endl;
  cout << "Eof: " << inp->eof() << endl;
  cout << "Pos: " << inp->tell() << endl;

  cout << "Entire stream from pointer: " << (char*)inp->getPtr() << endl;
  
  return 0;
}
