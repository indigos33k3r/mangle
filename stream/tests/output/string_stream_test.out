Size: 12
Pos: 0
Seeking...
Pos: 3
Reading: 4
Four bytes: lo w
Eof: 0
Pos: 7
Seeking again...
Pos: 12
Eof: 1
Seek to 6
Eof: 0
Pos: 6
Over-reading: 6
Result: world!
Eof: 1
Pos: 12
Finally, reading the entire string: 11
Result: hello world
Eof: 0
Pos: 11
Entire stream from pointer: hello world!
